gleambatch.py

```diff
@@ -0,0 +1,227 @@
+
+    Returns an updated GleamDef object.
+    """
+
+    # Fix any missing / inconsistent pops
+    rs.heuristic_set_pops()
+    rs.fix_min_pops()
+
+    # Load and apply CSSE
+    csse = CSSEData()
+    csse.load(batch.config["CSSE_dir"] + "/time_series_19-covid-{}.csv")
+    csse.apply_to_regions(rs)
+
+    if batch.config["use_foretold"]:
+        log.info("Loading and applying Foretold data")
+        # Load and apply FT
+        ft = FTData()
+        ft.load(batch.config["foretold_file"])
+        ft_before = datetime.datetime.combine(
+            batch.config["start_date"], datetime.time(23, 59, 59)
+        ).astimezone()
+        ft.apply_to_regions(rs, before=ft_before)
+
+    # TODO: This is asking for a good refactor of the redistribution code ...
+    override = batch.config.get("region_active_estimates")
+    if override is not None:
+        log.info(f"Overriding 'ft_mean' for {len(override)} regions ...")
+        for key, est in override.items():
+            rs[key].est["ft_mean"] = est
+
+    # Main computation: propagate estimates and fix for consistency with CSSE
+    FTData.propagate_down(rs)
+
+    # Propagate estimates upwards to super-regions
+    rs.fix_min_est(
+        "est_active", keep_nones=True, minimum_from="csse_active", minimum_mult=2.0
+    )  ## TODO: param for mult
+
+    rs.check_missing_estimates("est_active")
+
+    # Store the initial estimates in batch data
+    rs.fix_min_est("est_active")
+    batch.store_region_estimates(rs, "est_active", "FT_Infected")
+    # Finally, propagate csse upwards and also store
+    for loc_key, rem_key in [
+        ("csse_deaths", "JH_Deaths"),
+        ("csse_confirmed", "JH_Confirmed"),
+        ("csse_recovered", "JH_Recovered"),
+        ("csse_active", "JH_Infected"),
+    ]:
+        # Fix one-city provinces
+        for r in rs.regions:
+            if (
+                r.parent is not None
+                and len(r.parent.sub) == 1
+                and r.est.get(loc_key) is None
+            ):
+                r.est[loc_key] = r.parent.est.get(loc_key)
+        # Propagate up
+        rs.fix_min_est(loc_key)
+        # Store in batch
+        batch.store_region_estimates(rs, loc_key, rem_key)
+
+
+def estimates_to_gleamdef(batch, rs: Regions, input_xml_path):
+    gv = GleamDef(input_xml_path)
+    gv.set_start_date(batch.config["start_date"])
+    gv.clear_seeds()
+    for comp, coef in batch.config["compartments_mult"].items():
+        gv.add_seeds(rs, est_key="est_active", compartments={comp: coef})
+
+    return gv
+
+
+def parameterize(batch, gv):
+    for mit in batch.config["mitigations"]:
+        for sce in batch.config["scenarios"]:
+            gv2 = gv.copy()
+            gv2.set_seasonality(sce["param_seasonalityAlphaMin"])
+            gv2.set_traffic_occupancy(sce["param_occupancyRate"])
+            gv2.set_beta(mit["param_beta"])
+            gv2id = "{}.574".format(random.randint(1700000000000, 1800000000000))
```

Comment:
Huh, those are big numbers :grin: . What is it needed for? It's an ID which needs to be timestampy or? That .574 is there why?

Reply:
It is near the common gleamviz ID range and I chose that to be bit away from current gleamviz IDs .. but it IS timestampy :grin: Thanks for noticing, I will create it from timestamps now! 

Re `.574` - I don't know; all my gleamviz-created defs seemed to have it.

Reply:
> Re .574 - I don't know; all my gleamviz-created defs seemed to have it.

Rather not to change or it all breaks :smile: ...

Reply:
Thanks for noticing the timestapiness!

Reply:
(Now we will be much more sneaky in their DB :grin: )